packages/content-management/favorites/favorites_server/src/favorites_saved_object.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { schema } from '@kbn/config-schema';
+import type { SavedObjectsType } from '@kbn/core/server';
+
+export interface FavoritesSavedObjectAttributes {
+  userId: string;
+  favorites: Record<string, Array<{ id: string }>>;
+}
+
+const schemaV1 = schema.object({
+  userId: schema.string(),
+  favorites: schema.mapOf(
+    schema.string(), // object-type
+    schema.arrayOf(schema.object({ id: schema.string() })) // object id
```

Comment:
I did not test --- will this schema allow us to, given a set of Object IDs resolve which users favourited them? Kind of like what we do with SO references `packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/search/search_dsl/references_filter.ts`

Basically doing an inverted query SO ID => array of user IDs, I can imagine this being useful for certain content management UIs. Not a blocker, just curious if it was tested.

Reply:
Thanks! I haven't tested, no. But it anyway needs the data to be indexed, right? And I intentionally didn't add it to the mapping. 

I now wonder if I should simplify and put the data similar to saved object references 

favorites: [{id: string, type: string}]

Initially I used the grouping by type to do the partial updates and fetch only needed type using find api. but after learning that there is no actual partial updates with `so.update` now and using the custom id to rely on `get` there is not much benefit to the byType grouping.

Reply:
> I now wonder if I should simplify and put the data similar to saved object references
> favorites: [{id: string, type: string}]

I think if it provides the most querying flexibility then this makes sense to me!

Reply:
Sounds good, I simplified to an array

Reply:
> favorites: [{id: string, type: string}]

Does this really provide the most querying flexibility? Will we index them as `type: "nested"`?

Reply:
For now we don't index, because it is not needed. but if/when needed, yes, index as `type: "nested"`

... 


I am actually looking at the telemtry requirements. Here are the reqeuirements for the snapshot telemetry:

```
Snapshot telemetry
- Total starred dashboards per deployment
- Average starred dashboards per user for a given deployment
- Max starred dashboards per user for a given deployment
```

As I undetstand, to gather the telemtry, I'll already will have to add that `type: "nested"` to the mapping, right 

@afharo, do you have other suggestions for the structure? do you think that grouping by type is better / more flexibile? It seems it would be less flexible if we'd want to gather such telemetry for all types in total, non dependent on the type

Reply:
> I think we should differentiate the feature requirements and the telemetry requirements (which also makes the distinction between the reporting format and the indexing format).
> A custom usageCollector can shape the data however we want to ship it (even reformat the SO documents to [{id: string, type: string}] if we need to).
>I don't think I'm in a situation to suggest the best format needed for the feature. But, please, don't keep telemetry in mind when designing the data model. It can be retrieved and reshaped later :)

got it, thanks

> Maybe, thinking about consumption, it's best to group by type (or even have one document per type?). It all comes down to how we plan to query the SOs for the feature.

Current requirement is simple, it is enough that we just retrieve current user's favorites object and use it client side.
We might anticiapate something like "sort dashboard by most favorited" or "show users that favorites this object" but this isn't required today.

Initially I implemented the grouping by type inside an so, because I wanted to isolate the partially update to only the type that is currently being changed and fetch only the type being fetched. But after learning that there is no actual partial updates with `so.update` now and using the custom id to rely on `get` instead of `find` with `field` there is no much benefit to the by type grouping. So simplifing to array made sense.

Reply:
> Initially I implemented the grouping by type inside an so, because I wanted to isolate the partially update to only the type that is currently being changed and fetch only the type being fetched. But after learning that there is no actual partial updates with so.update now and using the custom id to rely on get instead of find with field there is no much benefit to the by type grouping. So simplifing to array made sense.

It makes perfect sense to me! My only concern with `[{id: string, type: string}]` is that, given `nested` is not supported by the SO API, we can only do `favorites.id: <ID 1> and favorites.type: <Type A>`, but it can yield any of the following results:

```
{ id: <ID 1>, type: <Type A> } // what we expect

// or these mixed results
{ id: <ID 1>, type: <Type B> }
{ id: <ID 2>, type: <Type A> }
```

Following the "partial updates" concept, does it make sense for the document to have the following structure?

```
{
  userId: string;
  soType: string; // e.g.: "dashboard"
  favorite_ids: string[]
}
```

Reply:
> given nested is not supported by the SO API

Good to know! Thanks for raising that constraint @afharo !